### PR TITLE
Implement redux on the form to add interactions

### PR DESCRIPTION
**Pull request by rafenden**
## Description of change

Add redux store to the new interaction form.

## Test instructions

1. Go to https://data-hub-fro-interactio-r1yy7f.herokuapp.com/interactions/create?company_id=5bb239e4-b227-4a84-bc2e-3420fc129447
2. Confirm if the form works the same as https://data-hub-fro-interactio-r1yy7f.herokuapp.com/companies/5bb239e4-b227-4a84-bc2e-3420fc129447/interactions/create

### CODE REVIEW

File: src/client/components/Form/index.jsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      {}
+    )
+
+const Form = ({ formAttributes, onSubmit, children, ...store }) => (
+  <useFormContext.Provider {...store}>
+    <form
+      {...formAttributes}
+      noValidate={true}
+      onSubmit={(e) => {
+        e.preventDefault()
```
Comment (peterhudec) [inline src/client/components/Form/index.jsx]: The blanket e.preventDefault() doesn't allow the form to be submitted the traditional way, but in this very PR there is such a usecase (see my comment about csrfToken).

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        options={contacts}
+        isMulti={true}
+      />
+
+      <FieldTypeahead
+        name="advisers"
+        label="Adviser(s)"
+        placeholder="-- Select adviser --"
+        required="Select at least one adviser"
+        options={advisers}
```
Comment (mforner13) [inline src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx]: Apologies as I know this is something I should have spotted in the previous PR, but in other places for advisers we've used the async version of the typeahead, so as not to be loading all 4,000ish of them into the DOM. Would this be a better option here than getting all advisers in the controller and passing them in?

File: src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import React from 'react'
+import { FieldRadios, useFormContext } from 'data-hub-components'
+import { THEMES, SERVICE_CONTEXTS } from '../../../constants'
+
+const getOnChangeHandler = (fieldName, setFieldValue) => (e) => {
+  setFieldValue('service', '')
+  setFieldValue(fieldName, e.target.value)
+}
```
[the remark this answers is not in the document]
Comment (mforner13) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx]: @peterhudec As you said I think they're used to conditionally show different versions of the form in StepInteractionDetails.  Are you saying that it as it's so tightly coupled with what gets displayed in the next step then the name setFieldValue should reflect this more?

[the remark this answers is not in the document]
Comment (peterhudec) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx]: @mforner13 The problem is that this component sets state it doesn't itself need and only does it for the sake of another component for which it is crucial. Any unit of code, be it a component, function or a class should never make any assumptions about the outside world, but only about their interface. If they do, it makes them tightly coupled to whatever they make assumptions about.

To be fair, it is possible and valid in Redux for a component to read state of another component, but one should never make a component set state only because some other component needs it.

File: src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx [hunk 4]
```diff
@@ -1,326 +1,33 @@
...
-const typeaheadOptionProp = PropTypes.shape({
-  label: PropTypes.string.isRequired,
-  value: PropTypes.string.isRequired,
-})
-
-const typeaheadOptionsListProp = PropTypes.arrayOf(typeaheadOptionProp)
+const AddInteractionForm = ({ csrfToken, ...interactionDetailsProps }) => (
+  <Form
+    id="interaction-form"
+    onSubmit={(values) => onSubmit(values, csrfToken)}
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx:15]: The submission is out of the scope for this PR, I removed `csrfToken` as suggested.

File: src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx [hunk 5]
```diff
@@ -1,326 +1,33 @@
...
-  )
-}
-
-const typeaheadOptionProp = PropTypes.shape({
-  label: PropTypes.string.isRequired,
-  value: PropTypes.string.isRequired,
-})
-
-const typeaheadOptionsListProp = PropTypes.arrayOf(typeaheadOptionProp)
+const AddInteractionForm = ({ csrfToken, ...interactionDetailsProps }) => (
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx:12]: It's because the props required on the second step (the interaction details step) were grouped together to make it clear that they aren't used anywhere else.

File: src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx [hunk 6]
```diff
@@ -1,326 +1,33 @@
...
+  <Form
+    id="interaction-form"
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx:14]: Agree, ID updated!

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  FieldTextarea,
+  FieldCheckboxes,
+} from 'data-hub-components'
+
+import {
+  THEMES,
+  SERVICE_CONTEXTS,
+  SERVICE_DELIVERY_STATUS_COMPLETED,
+} from '../../../constants'
+import PropTypes from 'prop-types'
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:20]: Good point, the code was updated!

File: src/client/components/Form/index.jsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+const componentStateToProps = (state) => ({
+  ...state,
+  getStepIndex: (stepName) => {
+    const index = state?.steps?.indexOf(stepName)
+    return index !== -1 ? index : null
+  },
+  isFirstStep: () => state?.currentStep === 0,
+  isLastStep: () => state?.currentStep === state?.steps?.length - 1,
+  getFieldState: (fieldName) => ({
+    value: state?.values?.[fieldName] ?? '',
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:73]: I want an empty string here and not `undefined` to make a form element a [controlled one](https://reactjs.org/docs/uncontrolled-components.html).

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+
+const getServiceContext = (values) => {
+  switch (values?.theme) {
+    case THEMES.EXPORT:
+      return values?.kind_export
+    case THEMES.INVESTMENT:
+      return SERVICE_CONTEXTS.INVESTMENT_INTERACTION
+    case THEMES.OTHER:
+      return values?.kind_other
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:29]: Good point, the code was updated.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  defaultAdviser,
+  services,
+  serviceDeliveryStatuses,
+  policyAreas,
+  policyIssueTypes,
+  communicationChannels,
+  countries,
+  activeEvents,
+}) => {
+  const { values } = useFormContext()
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:66]: The field components depends on the context so it has to stay for now.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+          }))}
+        />
+      ))}
+
+      {isServiceDelivery && isTapService(selectedService) && (
+        <>
+          <FieldSelect
+            label="Service status (optional)"
+            name="service_delivery_status"
+            emptyOption="-- Select service status --"
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:107]: Some existing forms on DH doesn't use dashes.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+  serviceDeliveryStatuses,
+  policyAreas,
+  policyIssueTypes,
+  communicationChannels,
+  countries,
+  activeEvents,
+}) => {
+  const { values } = useFormContext()
+
+  const today = moment()
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:68]: I replied above on your `useFormContext` question.

What is your suggestion on `moment()`?

File: src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+const StepInteractionType = () => {
+  const { setFieldValue } = useFormContext()
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx:11]: I think we discussed this offline.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionType.jsx:8]: This handler is needed to reset services field when the type of interaction is updated from the first step.

File: src/client/actions.js [hunk 14]
```diff
@@ -43,3 +43,17 @@ export const REFFERAL_LIST__LOADED = 'REFFERAL_LIST__LOADED'
...
 export const EXPORT_WINS__SELECT_PAGE = 'EXPORT_WINS__SELECT_PAGE'
+
+export const FORM__FIELD_SET_VALUE = 'FORM__FIELD_SET_VALUE'
+export const FORM__FIELD_TOUCHED = 'FORM__FIELD_TOUCHED'
+
+export const FORM__FIELD_REGISTER = 'FORM__FIELD_REGISTER'
+export const FORM__FIELD_DEREGISTER = 'FORM__FIELD_DEREGISTER'
+
+export const FORM__STEP_REGISTER = 'FORM__STEP_REGISTER'
+export const FORM__STEP_DEREGISTER = 'FORM__STEP_DEREGISTER'
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/actions.js:54]: To have dynamically added field (with validations).
We discussed this offline.

File: src/client/components/Form/index.jsx [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+import multiInstance from '../../utils/multiinstance'
+import reducer from './reducer'
+
+const validateForm = (state) =>
+  Object.values(state.fields)
+    .map((field) => ({
+      name: field.name,
+      error: castArray(field.validate)
+        .map((validator) => validator(state.values?.[field.name], field, state))
+        .filter((e) => e)[0],
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:26]: Having multiple validators is useful in the following scenerio:
```
const validateRequired = (value) => !value ? 'Date is required' : null
const validateDateFormat = (value) => !isValidDate(value) ? 'Date is invalid' : null

<FieldInput>
  name="date"
  validators={[
    validateRequired,
    validateDateFormat,
  ]}
/>

[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:26]: The filter() was updated as suggested, thanks!

File: src/client/components/Form/index.jsx [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    }))
+    .filter(({ error }) => error)
+    .reduce(
+      (prev, { name, error }) => ({
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:30]: Good point, the var was renamed!

File: src/client/components/Form/index.jsx [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+  FORM__VALIDATE,
+} from '../../../client/actions'
+import multiInstance from '../../utils/multiinstance'
+import reducer from './reducer'
+
+const validateForm = (state) =>
+  Object.values(state.fields)
+    .map((field) => ({
+      name: field.name,
+      error: castArray(field.validate)
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:24]: Good point! The code was updated!

File: src/client/components/Form/index.jsx [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        },
+        touched: {
+          ...prev.touched,
+          [name]: true,
+        },
+      }),
+      {}
+    )
+
+const Form = ({ formAttributes, onSubmit, children, ...store }) => (
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:43]: Thanks, the code was updated!

File: src/client/components/Form/index.jsx [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+const componentStateToProps = (state) => ({
+  ...state,
+  getStepIndex: (stepName) => {
+    const index = state?.steps?.indexOf(stepName)
+    return index !== -1 ? index : null
+  },
+  isFirstStep: () => state?.currentStep === 0,
+  isLastStep: () => state?.currentStep === state?.steps?.length - 1,
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:71]: They need to be (for now) as the existing `Step` component use them.

File: src/client/components/Form/index.jsx [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    </form>
+  </useFormContext.Provider>
+)
+
+const componentStateToProps = (state) => ({
+  ...state,
+  getStepIndex: (stepName) => {
+    const index = state?.steps?.indexOf(stepName)
+    return index !== -1 ? index : null
+  },
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:69]: `getStepIndex` is used by the `Step` component so it has to stay there (for now).

File: src/client/components/Form/reducer.js [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+        values: {
+          ...state.previousValues,
+          ...state.values,
+          [action.field.name]:
+            state.values?.[action.field.name] ?? action.field.initialValue,
+        },
+        touched: {
+          ...state.touched,
+          [action.field.name]: false,
+        },
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/reducer.js:39]: React form elements care, it was added to send a real value instead of `meta={{ touched: true }}`.

File: src/client/components/Form/index.jsx [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+  FORM__FIELD_TOUCHED,
+  FORM__FORWARD,
+  FORM__STEP_DEREGISTER,
+  FORM__STEP_REGISTER,
+  FORM__VALIDATE,
+} from '../../../client/actions'
+import multiInstance from '../../utils/multiinstance'
+import reducer from './reducer'
+
+const validateForm = (state) =>
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:20]: The `state` var wasn't destructed because it was passed to the validation function.

File: src/client/components/Form/index.jsx [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+    const index = state?.steps?.indexOf(stepName)
+    return index !== -1 ? index : null
+  },
+  isFirstStep: () => state?.currentStep === 0,
+  isLastStep: () => state?.currentStep === state?.steps?.length - 1,
+  getFieldState: (fieldName) => ({
+    value: state?.values?.[fieldName] ?? '',
+    touched: state?.touched?.[fieldName] ?? false,
+    error: state?.errors?.[fieldName],
+  }),
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:76]: As above, the `state` was spread and for that reason it wasn't destructed in the function signature.

File: src/client/components/Form/reducer.js [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+      }
+    case FORM__FIELD_DEREGISTER:
+      return {
+        ...state,
+        values: omit(state.values, action.fieldName),
+        errors: omit(state.errors, action.fieldName),
+        touched: omit(state.touched, action.fieldName),
+        fields: omit(state.fields, action.fieldName),
+      }
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/reducer.js:52]: We discussed that online, field and step registry was required to mound dynamic field with validation.

File: src/client/components/Form/reducer.js [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+        ...state,
+        steps: !state.steps.includes(action.stepName)
+          ? [...state.steps, action.stepName]
+          : state.steps,
+      }
+    case FORM__STEP_DEREGISTER:
+      return {
+        ...state,
+        steps: state.steps.filter((s) => s !== action.stepName),
+      }
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/reducer.js:104]: See my answer above.

File: src/client/components/Form/reducer.js [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+        },
+      }
+    case FORM__FIELD_TOUCHED:
+      return {
+        ...state,
+        touched: {
+          ...state.touched,
+          [action.fieldName]: true,
+        },
+      }
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/reducer.js:74]: See my answer above.

File: src/client/components/Form/index.jsx (hunk 1 above)
Comment (rafenden) [reply src/client/components/Form/index.jsx:49]: We shouldn't submit any forms the traditional way if we aim for SPA.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx (hunk 2 above)
Comment (rafenden) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:142]: Thanks, Miriam, this will be added in the next PR as I already reach the line limit.

File: src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx [hunk 27]
```diff
@@ -1,327 +1,27 @@
...
-  serviceDeliveryStatuses: typeaheadOptionsListProp.isRequired,
-  policyAreas: typeaheadOptionsListProp.isRequired,
-  policyIssueTypes: typeaheadOptionsListProp.isRequired,
-  communicationChannels: typeaheadOptionsListProp.isRequired,
-  countries: typeaheadOptionsListProp.isRequired,
-  activeEvents: typeaheadOptionsListProp.isRequired,
-  csrfToken: PropTypes.string.isRequired,
-}
+const AddInteractionForm = (props) => (
+  <Form id="add-interaction-form" onSubmit={(values) => onSubmit(values)}>
```
Comment (peterhudec) [inline src/apps/interactions/apps/add-interaction/client/AddInteractionForm.jsx:12]: We should not use IDs.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:68]: I'm afraid there's no way around it without a saga. You could use a `Task` for it, but that'd be an overkill.

File: src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/apps/interactions/apps/add-interaction/client/StepInteractionDetails.jsx:107]: Should they be used here then? We should be consistent with it.

File: src/client/components/Form/index.jsx (hunk 1 above)
Comment (peterhudec) [reply src/client/components/Form/index.jsx:49]: We hade a couple of features for which it would have been an overkill to submit the form in any other but the traditional way. We want to keep this option until we have a full SPA.

File: src/client/components/Form/index.jsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/client/components/Form/index.jsx:69]: That's fine, it just should not be defined here, but inside `Form` or outside of it, in which case it should take the `stepName` as another parameter.

File: src/client/components/Form/index.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/client/components/Form/index.jsx:71]: I see, they use the `state` from the closure. It suffers from the same problem as `getStepIndex` then. The problem is that the stateless `Form` you pass to `connect` or `multiinstance` requires four props which are statefull and not pure functions and should actually be in sync with each other. That makes the `Form` interface extremely complicated, hard to explain and satisfy and thus tightly coupled to `componentStateToProps`. The solution is as simple as moving all of them to the `Form` where you have access to the `state` (which shouldn't really be called like that).

File: src/client/components/Form/index.jsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/client/components/Form/index.jsx:76]: I didn't mean that. What I meant is that the `Form` requires a `getFieldState` function which doesn't take any arguments, but yet returns varying values i.e. it is stateful. To be precise, the `Form`'s props need to be are an extension of an instance of a stateful object which has the `getStepIndex`, `isFirstStep`, `isLastStep` and `getFieldState` methods, which are expected to work in conjunction. Although the `Form` is a stateless component, it requires a stateful combination of props. This inothing but OOP. 

You could say that the stateless `Form` will never be used anywhere else apart from being passed to `connect` but the problem is that a portion of it's logic is completely unnecessarily shoved into `componentStateToProps` which will be extremely hard to follow. Especially when what you expect to see passed to `connect` is a simple stateless component.

File: src/client/components/Form/reducer.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (peterhudec) [reply src/client/components/Form/reducer.js:39]: But `meta.touched` only needs a real value for the sake of inline validation before submit, which we agreed never to do, so we only ever need to have `meta.touched` hardcoded to `true`, which means that this logic is redundant and should go.

File: src/client/components/Form/index.jsx (hunk 1 above)
Comment (rafenden) [reply src/client/components/Form/index.jsx:49]: In either case you will have to create an endpoint to which the form will be posted so the amount of work would be the same therefore I'm not sure where is the overkill you are referring to.

File: src/client/components/Form/index.jsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (rafenden) [reply src/client/components/Form/index.jsx:69]: Good point, these function were moved to the `<Form>`

**Comment by paulgain** [comment]
**Looks really good Raf.**

There a couple of cosmetic issues that need addressing:

1. The components should be indented along with a vertical grey bar when conditionally revealing content. This applies to the other steps in this form as well:
https://design-system.service.gov.uk/components/radios/conditional-reveal/index.html 
2. The question to each component should be bold.

changes-required

**Comment by rafenden** [comment]
@paulgain 

> There a couple of cosmetic issues that need addressing

Ad. 1
GDS advise avoiding conditional reveal in a few cases.
> Keep it simple. If you need to add a lot of content, consider showing it on the next page in the process instead.
> Do not use this component to add conditionally revealing content to inline radios.
https://design-system.service.gov.uk/components/radios/#conditionally-revealing-content

Ad. 2
All the new React forms, govuk-react and GDS are using non-bolded labels and we should too.

**Comment by paulgain** [comment]
> Do not use this component to add conditionally revealing content to inline radios.

In my opinion this is ambiguous. To me, this means:
Do not use this component (i.e. radio buttons) that conditionally reveal (more) inline radio buttons
